File: effective/src/com/lab/java/zayson/item19/Person.java
package com.lab.java.zayson.item19;

public class Person {
    int distance;
    private final String name;
    private final int age;

    public Person(String name, int age) {
        distance = 0;
        this.name = name;
        this.age = age;
    }

    public void walk() {
        distance ++;
    }

    // 상위 클래스에서는 재정의한 메서드를 호출하지 않고 헬퍼 메서드로 추출
    public void stop() {
        // 자기 사용 메서드
        String record = recordHelper();
        System.out.println("record = " + record);
    }

    // private 도우미 메서드로 추출 (현재 클래스에서만 호출 가능)
    private String recordHelper() {
        System.out.println("기록 시작");
        return String.format("이름 : %s, 나이 ; %d, 이동 거리 ; %d", name, age, distance);
    }

    // 하위 클래스에서 재정의 가능한 메서드
    public String record() {
        return "";
    }
}

File: practice/src/main/java/zayson/java/lab/kotlin/grammer/lecture12/Person.kt
package zayson.java.lab.kotlin.grammer.lecture12


class Person private constructor(
    var name: String,
    var age: Int
) {
    // 코틀린은 static 키워드가 없음 -> companion object 사용
    companion object Factory : Log { // companion object의 이름 = Factory, Log인터페이스 상속
        private const val MIN_AGE = 1

        @JvmStatic // 자바 코드에서 코틀린 파일의 메서드를 Static 함수처럼 사용 가능
        fun newBaby(name: String): Person {
            return Person(name, MIN_AGE)
        }

        override fun log() {
            println("logging factory")
        }
    }
}

// 싱글톤 클래스
object Singleton {
    var a: Int = 0
}

fun main() {
    // 익명 클래스 사용 방법 -> object : 인터페이스명
    moveSomething(object : Movable {
        override fun fly() {
            println("날다")
        }

        override fun move() {
            println("움직이다")
        }
    })
}

private fun moveSomething(movable: Movable) {
    movable.move()
    movable.fly()
}